include/mbedtls/x509_crt.h

```diff
@@ -675,6 +746,232 @@ void mbedtls_x509_crt_restart_init( mbedtls_x509_crt_restart_ctx *ctx );
  */
 void mbedtls_x509_crt_restart_free( mbedtls_x509_crt_restart_ctx *ctx );
 #endif /* MBEDTLS_ECDSA_C && MBEDTLS_ECP_RESTARTABLE */
+
+/**
+ * \brief           Request CRT frame giving access to basic CRT fields
+ *                  and raw ASN.1 data of complex fields.
+ *
+ * \param crt       The CRT to use. This must be initialized and setup.
+ * \param dst       The address of the destination frame structure.
+ *                  This need not be initialized.
+ *
+ * \note            mbedtls_x509_crt_frame does not contain dynamic
+ *                  references and hence need not be freed. Users may
+ *                  e.g. allocate an instance of mbedtls_x509_crt_frame
+ *                  on the stack and call this function on it, in which
+ *                  case no allocation/freeing has to be done.
+ *
+ * \return          \c 0 on success. In this case, \p dst is updated
+ *                  to hold the frame for the given CRT.
+ * \return          A negative error code on failure.
+ */
+int mbedtls_x509_crt_get_frame( mbedtls_x509_crt const *crt,
+                                mbedtls_x509_crt_frame *dst );
+
+/**
+ * \brief           Request a PK context for the public key within a CRT.
+ *
+ * \param crt       The CRT to use. This must be initialized and setup.
+ * \param pk        The address of the destination PK context to fill.
+ *                  This must be initialized via mbedtls_pk_init().
+ *
+ * \return          \c 0 on success. In this case, the user takes ownership
+ *                  of the destination PK context, and is responsible for
+ *                  calling mbedtls_pk_free() on it once it's no longer needed.
+ * \return          A negative error code on failure.
+ */
+int mbedtls_x509_crt_get_pk( mbedtls_x509_crt const *crt,
+                             mbedtls_pk_context *pk );
+
+/**
+ * \brief           Request the subject name of a CRT, presented
+ *                  as a dynamically allocated linked list.
+ *
+ * \param crt       The CRT to use. This must be initialized and setup.
+ * \param subject   The address at which to store the address of the
+ *                  first entry of the generated linked list holding
+ *                  the subject name.
+ *
+ * \note            Depending in your use case, consider using the raw ASN.1
```

Comment:
@andresag01 Yes, `mbedtls_x509_crt_frame` is meant to be accessed directly. If it was C++, it would be all-public. 

> Also, I would guess that people rarely want to access these raw fields, otherwise they might as well just use the ASN.1 API directly and parse most of the certificate themselves.

Previously there were heap-allocated linked lists for the raw components of the extensions; now, we're providing the raw bounds of the extensions plus ASN.1 functions to traverse these. Together, this gives the same depth of parsing for the user. If the need more, they need to add it themselves or query for a feature - that's what it is before and after this PR. Looking at the documentation, though, it should mention the respective ASN.1 functions - one of which is currently still marked internal out of hesitance, and I'll remove that marking.

In general, the choice of what belongs to `mbedtls_x509_crt_frame` and what doesn't is up to debate - there is no god-given definition which would uniquely specify it, and I've changed my mind on it many times. For example, why are we having the raw bounds of some extensions, instead of just the raw bounds of the entire extension block? One has to find a balance between compactness of structure and code on the one hand, and both convenience and coverage of the essential functionality on the other hand, and the current version of `mbedtls_x509_crt_frame` is the best I could come up with so far - but feel free to convince me that it should look differently.

Reply:
Added references to `mbedtls_asn1_traverse_sequence_of()` in 3f30a33, and marking the latter as public but experimental in af4b933.